### PR TITLE
Draft: Docker Executor job isolation

### PR DESCRIPTION
**Pull request by firecow**
- Fixed bug with extends field, when string value was used in multi level job extends
- Stopped printing stage's information, print next to job instead
- Add docker executor job isolation via rsync

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -19,13 +19,15 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
       - name: Install modules
         run: npm install
+      - name: Generate SSH key
+        run: cat /dev/zero | ssh-keygen -q -N ""
       - name: Run tests
-        run: npm run test
+        run: GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" npm --silent run test
```
Comment (moberghammer) [inline .github/workflows/build.yml:33]: Hmm hvorfor ignore knownhosts

File: src/job.ts [hunk 2]
```diff
@@ -255,99 +200,97 @@ export class Job {
...
-        await this.copyArtifactsToHost();
-        await this.removeContainer(this.containerId);
-
         process.stdout.write(`${this.getFinishedString(startTime)}\n`);
 
         this.running = false;
         this.finished = true;
 
+        if (this.containerId) {
+            await Utils.spawn(`docker rm -f ${this.containerId} > /dev/null`);
```
Comment (moberghammer) [inline src/job.ts:209]: Kunne godt være mere DRY er gentaget 3 x

File: src/job.ts [hunk 3]
```diff
@@ -255,99 +200,97 @@ export class Job {
...
+            dockerCmd += `\texit 1\n`;
+            dockerCmd += `fi\n"`
+            const {stdout: containerId} = await Utils.spawn(dockerCmd, this.cwd, { ...process.env, ...this.expandedVariables, });
+            this.containerId = containerId.replace("\n", "");
+        }
 
-            for (const [key, value] of Object.entries(this.expandedVariables)) {
-                await fs.appendFile(entrypointPath, `export ${key}="${String(value).trim()}"\n`);
-            }
+        const cp = childProcess.spawn(this.containerId ? `docker attach ${this.containerId}` : `bash`, {
```
Comment (moberghammer) [inline src/job.ts:265]: du kan vel ikke regne med at alle images har bash

File: src/parser.ts [hunk 4]
```diff
@@ -48,8 +51,8 @@ export class Parser {
         let gitlabUserEmail, gitlabUserName;
 
         try {
-            const res = await Utils.spawn(`git config user.email`, {shell: true, cwd});
-            gitlabUserEmail = res.trimEnd();
+            const {stdout: gitConfigEmail} = await Utils.spawn(`git config user.email`, cwd);
+            gitlabUserEmail = gitConfigEmail.trimEnd();
         } catch (e) {
             // process.stderr.write(`${yellow("git config user.email is undefined, defaulting to `local@gitlab.com`")}`);
```
Comment (moberghammer) [inline src/parser.ts:57]: dead code

File: src/tests/parser.test.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// import {Parser} from "../parser";
```
Comment (moberghammer) [inline src/tests/parser.test.ts:1]: dead code

File: src/tests/spawn.test.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import {Utils} from "../utils";
+
+// test('Spawn successful command', async() => {
```
Comment (moberghammer) [inline src/tests/spawn.test.ts:3]: dead code

File: src/job.ts (hunk 3 above)
Comment (firecow) [reply src/job.ts:265]: Det er hosten jeg regner med har bash...

**Review by moberghammer — APPROVED** [review]
LGTM

**Pull request merged**